File: apps/android/app/src/main/kotlin/app/pbbls/android/services/PebbleDraftsService.kt
package app.pbbls.android.services

import androidx.compose.runtime.staticCompositionLocalOf
import app.pbbls.android.features.path.models.OffsetDateTimeSerializer
import app.pbbls.android.features.path.models.PebbleDraftPayload
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.time.OffsetDateTime
import java.util.UUID

/** A `pebble_drafts` row as the drafts list needs it. */
@Serializable
data class PebbleDraftRecord(
    val id: String,
    val payload: PebbleDraftPayload,
    @SerialName("updated_at")
    @Serializable(with = OffsetDateTimeSerializer::class)
    val updatedAt: OffsetDateTime,
)

/**
 * Server-side drafts (M47) — ports iOS `PebbleDraftsService.swift`.
 *
 * Direct RLS-scoped single-table calls (design D6, the sanctioned cross-surface
 * pattern): nothing here is multi-table and nothing emits karma. Errors
 * propagate to the caller, which owns loading/error view state.
 *
 * Zero karma is structural rather than enforced: `create_pebble` is the schema's
 * only `pebble_created` emitter and a draft never reaches it.
 */
class PebbleDraftsService(
    private val supabase: SupabaseService,
) {
    private companion object {
        const val TABLE = "pebble_drafts"
        const val COLUMNS = "id, payload, updated_at"
    }

    /**
     * Most recently saved first. `updated_at` is trigger-maintained server-side,
     * so the ordering does not depend on device clocks.
     */
    suspend fun list(): List<PebbleDraftRecord> =
        supabase.client
            .from(TABLE)
            .select(Columns.raw(COLUMNS)) {
                order("updated_at", Order.DESCENDING)
            }.decodeList()

    suspend fun load(id: String): PebbleDraftRecord? =
        supabase.client
            .from(TABLE)
            .select(Columns.raw(COLUMNS)) {
                filter { eq("id", id) }
                limit(1)
            }.decodeList<PebbleDraftRecord>()
            .firstOrNull()

    /**
     * Upsert. Passing an existing [id] replaces that draft's payload wholesale —
     * the point of the jsonb column, since a coalescing update could never clear
     * a field the user emptied.
     *
     * `user_id` is explicit because the RLS `with check` compares it to
     * `auth.uid()`.
     */
    suspend fun save(
        payload: PebbleDraftPayload,
        id: String?,
        userId: String,
    ): String {
        val rowId = id ?: UUID.randomUUID().toString()
        supabase.client
            .from(TABLE)
            .upsert(DraftUpsertPayload(id = rowId, userId = userId, payload = payload))
        return rowId
    }

    suspend fun delete(id: String) {
        supabase.client
            .from(TABLE)
            .delete {
                filter { eq("id", id) }
            }
    }

    /**
     * Count for the Path entry point. Selects `id` alone rather than reusing
     * [list]: this runs on every Path composition, and pulling every draft's full
     * jsonb payload to render one number is waste on the app's home screen.
     */
    suspend fun count(): Int =
        supabase.client
            .from(TABLE)
            .select(Columns.raw("id"))
            .decodeList<DraftIdRow>()
            .size

    /**
     * Whether a resumed draft's glyph is still usable (own ∪ system ∪ entitled).
     * Lives here rather than on `GlyphService` because draft hydration is its only
     * caller: `can_use_glyph` is the predicate `create_pebble` enforces, so
     * checking it while hydrating means publishing cannot fail on 42501 later
     * (design D7).
     */
    suspend fun canUseGlyph(
        glyphId: String,
        userId: String,
    ): Boolean =
        supabase.client.postgrest
            .rpc(
                "can_use_glyph",
                buildJsonObject {
                    put("p_glyph_id", glyphId)
                    put("p_user", userId)
                },
            ).decodeAs()
}

/**
 * `user_id` is explicitly encoded so the RLS `with check` sees it; `payload`
 * rides along as the nested jsonb.
 */
/** Minimal projection for the count query. */
@Serializable
private data class DraftIdRow(
    val id: String,
)

@Serializable
private data class DraftUpsertPayload(
    val id: String,
    @SerialName("user_id")
    val userId: String,
    val payload: PebbleDraftPayload,
)

val LocalPebbleDraftsService =
    staticCompositionLocalOf<PebbleDraftsService> {
        error("LocalPebbleDraftsService not provided — wrap the tree in MainActivity's CompositionLocalProvider")
    }

File: apps/ios/Pebbles/Services/PebbleDraftsService.swift
import Foundation
import Observation
import Supabase
import os

/// A `pebble_drafts` row as the drafts list needs it.
///
/// `updated_at` is decoded from its **string** form through `ISO8601Flexible`
/// rather than left to whatever date strategy the ambient decoder happens to
/// carry (#649). Postgres returns microsecond precision, and a decoder expecting
/// a numeric or fixed-precision date fails the whole row — which would break the
/// drafts list outright, not just the timestamp.
struct PebbleDraftRecord: Identifiable, Decodable, Equatable {
    let id: UUID
    let payload: PebbleDraftPayload
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case payload
        case updatedAt = "updated_at"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        payload = try container.decode(PebbleDraftPayload.self, forKey: .payload)
        let raw = try container.decode(String.self, forKey: .updatedAt)
        guard let parsed = ISO8601Flexible.parse(raw) else {
            throw DecodingError.dataCorruptedError(
                forKey: .updatedAt, in: container,
                debugDescription: "unparseable updated_at: \(raw)"
            )
        }
        updatedAt = parsed
    }

    /// Memberwise init, for tests and previews.
    init(id: UUID, payload: PebbleDraftPayload, updatedAt: Date) {
        self.id = id
        self.payload = payload
        self.updatedAt = updatedAt
    }
}

/// Server-side drafts (M47). Direct single-table CRUD on `pebble_drafts`,
/// owner-scoped by RLS — the sanctioned direct-client pattern, since nothing
/// here is multi-table and nothing emits karma.
///
/// Zero karma is structural rather than enforced: `create_pebble` is the
/// schema's only `pebble_created` emitter and a draft never reaches it.
///
/// Production wiring lives in `PebblesApp`, alongside `SnapURLCache`.
@Observable
@MainActor
final class PebbleDraftsService {

    private let client: SupabaseClient
    private let logger = Logger(subsystem: "app.pbbls.ios", category: "pebble-drafts")

    /// Mirrors the count shown on the Path entry point. Refreshed by `list()`.
    private(set) var count: Int = 0

    init(client: SupabaseClient) {
        self.client = client
    }

    /// Most recently saved first — `updated_at` is trigger-maintained server-side,
    /// so this ordering does not depend on client clocks.
    func list() async throws -> [PebbleDraftRecord] {
        let rows: [PebbleDraftRecord] = try await client
            .from("pebble_drafts")
            .select("id, payload, updated_at")
            .order("updated_at", ascending: false)
            .execute()
            .value
        count = rows.count
        return rows
    }

    func load(id: UUID) async throws -> PebbleDraftRecord? {
        let rows: [PebbleDraftRecord] = try await client
            .from("pebble_drafts")
            .select("id, payload, updated_at")
            .eq("id", value: id)
            .limit(1)
            .execute()
            .value
        return rows.first
    }

    /// Upsert. `id` nil creates; passing one replaces that draft's payload
    /// wholesale — which is the point of the jsonb column, since a coalescing
    /// update could never clear a field the user emptied.
    ///
    /// `userId` is explicit because the RLS `with check` compares it to
    /// `auth.uid()`.
    @discardableResult
    func save(payload: PebbleDraftPayload, id: UUID?, userId: UUID) async throws -> UUID {
        let row = DraftUpsertPayload(id: id ?? UUID(), userId: userId, payload: payload)
        try await client
            .from("pebble_drafts")
            .upsert(row, onConflict: "id")
            .execute()
        return row.id
    }

    func delete(id: UUID) async throws {
        try await client
            .from("pebble_drafts")
            .delete()
            .eq("id", value: id)
            .execute()
    }

    /// Best-effort cleanup after a successful publish: the pebble exists, so a
    /// failure here must not surface as a publish failure. Worst case a stale
    /// draft is left in the list.
    func deleteIgnoringFailure(id: UUID) async {
        do {
            try await delete(id: id)
        } catch {
            logger.warning("draft cleanup after publish failed: \(error.localizedDescription, privacy: .private)")
        }
    }

    /// Refresh `count` without surfacing an error — it only drives an entry
    /// point's badge, and a missing badge beats an error state on the Path.
    ///
    /// Selects `id` alone rather than reusing `list()`: this runs on every Path
    /// appearance, and pulling every draft's full jsonb payload to render one
    /// number is waste on the app's home screen.
    func refreshCount() async {
        struct IdRow: Decodable { let id: UUID }
        do {
            let rows: [IdRow] = try await client
                .from("pebble_drafts")
                .select("id")
                .execute()
                .value
            count = rows.count
        } catch {
            logger.error("draft count refresh failed: \(error.localizedDescription, privacy: .private)")
        }
    }

    /// `user_id` is explicitly encoded so the RLS `with check` sees it; `payload`
    /// rides along as the nested jsonb.
    private struct DraftUpsertPayload: Encodable {
        let id: UUID
        let userId: UUID
        let payload: PebbleDraftPayload

        enum CodingKeys: String, CodingKey {
            case id
            case userId = "user_id"
            case payload
        }
    }
}
